### PR TITLE
Wrap bank keeper for EVM to convert decimals

### PR DESCRIPTION
**Pull request by drklee3**
Sending transactions seem to work properly with `ukava` as the evm_denom now.

### CODE REVIEW

File: app/evm_keeper.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+// EVMBankKeeper is a wrapper for bank keeper that converts between EVM and native decimals
+type EVMBankKeeper struct {
+	bankKeeper bankkeeper.Keeper
+}
+
+var _ evmtypes.BankKeeper = (*EVMBankKeeper)(nil)
+
+// NewEVMBankKeeper returns a wrapped bank keeper that converts between EVM and native decimals
+func NewEVMBankKeeper(bk bankkeeper.Keeper) EVMBankKeeper {
```
Comment (nddeluca) [inline app/evm_keeper.go:21]: It'd be good to unit test all of the functions here -- table tests with specific examples would be great for a wrapper like this.

File: app/evm_keeper.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	coin.Amount = coin.Amount.Mul(conversionMultiplier)
+
+	return coin
+}
+
+// convertCoinFromEvm converts an EVM sdk.Coin with 18 decimals back to native 6 decimal sdk.Coin
+func convertCoinFromEvm(coin sdk.Coin) sdk.Coin {
+	coin.Amount = coin.Amount.Quo(conversionMultiplier)
+
+	if coin.IsZero() {
```
Comment (nddeluca) [inline app/evm_keeper.go:68]: I *think* we actually want to truncate to zero here and return an error from any method converting from evm 18 decimals to 6 decimals.

Even though it's small, rounding up ends up operating on more coins than specified -- which should be avoided.  An explicit error is probably best?

Comment (nddeluca) [reply app/evm_keeper.go:68]: The edge cases also make great test examples -- most conversion code is much easier to understand from the tests in addition to having the implementation.

File: app/evm_keeper.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		coin.Amount = sdk.OneInt()
+	}
+
+	return coin
+}
+
+// convertCoinFromEvm converts EVM sdk.Coins with 18 decimals back to native 6 decimal sdk.Coins
+func convertCoinsFromEvm(coins sdk.Coins) sdk.Coins {
+	// Use a new sdk.Coins, do NOT modify in place as it is used afterwards
```
Comment (nddeluca) [inline app/evm_keeper.go:77]: This type of behavior tends to get lost over time, if it's possible lets add a test case that ensures we do not modify in place.

File: app/evm_keeper.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	evmtypes "github.com/tharsis/ethermint/x/evm/types"
+)
+
+// Convertion between native gas KAVA (6) to EVM (18).
+// 12 decimal difference, so 1_000_000_000_000.
+var conversionMultiplier = sdk.NewInt(1_000_000_000_000)
+
+// EVMBankKeeper is a wrapper for bank keeper that converts between EVM and native decimals
+type EVMBankKeeper struct {
+	bankKeeper bankkeeper.Keeper
```
Comment (nddeluca) [inline app/evm_keeper.go:15]: Since we intend to only use with the EVM -- we could also depend the abstract https://github.com/tharsis/ethermint/blob/main/x/evm/types/interfaces.go#L29.  Also, it's usually preferable to depend on interfaces where we can (and if it doesn't introduce unnecessary complexity).

File: app/evm_keeper.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+// convertCoinToEvm converts a sdk.Coin with native decimals to an sdk.Coin with evm decimals
+func convertCoinToEvm(coin sdk.Coin) sdk.Coin {
+	coin.Amount = coin.Amount.Mul(conversionMultiplier)
+
+	return coin
+}
+
+// convertCoinFromEvm converts an EVM sdk.Coin with 18 decimals back to native 6 decimal sdk.Coin
+func convertCoinFromEvm(coin sdk.Coin) sdk.Coin {
```
Comment (nddeluca) [inline app/evm_keeper.go:65]: I like the unexported helpers here 👍

File: app/evm_keeper.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+var _ evmtypes.BankKeeper = (*EVMBankKeeper)(nil)
+
+// NewEVMBankKeeper returns a wrapped bank keeper that converts between EVM and native decimals
+func NewEVMBankKeeper(bk bankkeeper.Keeper) EVMBankKeeper {
+	return EVMBankKeeper{
+		bankKeeper: bk,
+	}
+}
+
+func (bk EVMBankKeeper) GetBalance(ctx sdk.Context, addr sdk.AccAddress, denom string) sdk.Coin {
```
Comment (nddeluca) [inline app/evm_keeper.go:27]: Could add doc comments for the exported functions

File: app/evm_keeper.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+// Convertion between native gas KAVA (6) to EVM (18).
+// 12 decimal difference, so 1_000_000_000_000.
+var conversionMultiplier = sdk.NewInt(1_000_000_000_000)
+
+// EVMBankKeeper is a wrapper for bank keeper that converts between EVM and native decimals
+type EVMBankKeeper struct {
+	bankKeeper bankkeeper.Keeper
+}
+
+var _ evmtypes.BankKeeper = (*EVMBankKeeper)(nil)
```
Comment (nddeluca) [inline app/evm_keeper.go:18]: Nice, like these types of assertions

**Review by nddeluca — CHANGES_REQUESTED** [review]
Looks great.

Just two things:
  - Let's add full test coverage -- helps clarity and maintainability.
  - I think we should avoid rounding small amounts up to 1

File: app/evm_keeper.go (hunk 2 above)
Comment (drklee3) [reply app/evm_keeper.go:77]: Yeah, having an error here instead does makes sense. It should only be zero if it's dealing with values less than `1ukava` which shouldn't be valid anyways I suppose. Related thing I need to also check is if typical gas fees via wallets like MetaMask are less than `1ukava` which doesn't seem like great UX if users need to manually adjust it each time to not have it error.

Will add more tests for this after the weekend

Comment (drklee3) [reply app/evm_keeper.go]: I think this might need specific logic to handle gas fees or some modification to fee calculations possibly in Ethermint, as some test transactions via MetaMask had default gas fees < 1ukava which causes transactions to fail with a panic.

**Review by DracoLi — COMMENTED** [review]
do we want to return only the spendable balance for the EVM to account for vesting accounts?

**Comment by drklee3** [comment]
> do we want to return only the spendable balance for the EVM to account for vesting accounts?

Right, added now. I left the test case to check vesting account balances as a TODO since EthAccount and vesting accounts aren't compatible right now

**Review by DracoLi — APPROVED** [review]
